refactor: maintainability improvements

carlocorradini opened the pull request:
This PR makes a few changes to the repository to make it easier to manage and/or utilize generally:

## Changes

- Added [Dependabot](https://github.com/features/security)
   _Automatically monitor and fix vulnerable dependencies_
- Added [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)
   _File to define individuals or teams that are responsible for code in a repository_
- Better CI/CD (see `.github/workflows`)
- Added [clippy](https://doc.rust-lang.org/clippy)
   A collection of lints to catch common mistakes and improve your Rust code.
   See `lints` in `Cargo.toml`
   Some code has been refactored (**no breaking changes**) to fix all lints
- Added [.vscode](https://code.visualstudio.com) directory
   - Added [extensions.json](https://code.visualstudio.com/docs/editor/extension-marketplace#_workspace-recommended-extensions) file
      _Workspace recommended extensions_
   - Added [settings.json](https://code.visualstudio.com/docs/getstarted/settings#_settingsjson)
      _Workspace settings_
- Added [EditorConfig](https://editorconfig.org)
   _maintain consistent coding styles for multiple developers working on the same project across various editors and IDEs_
- Added [.gitattributes](https://git-scm.com/docs/gitattributes)
   _Defining attributes per path_
- Better `.gitignore`
- `Cargo.lock` should be ignored in libraries (added in `.gitignore`)
- Fixed some typos
- Better `README.md` (all `*.md` files in general) with better badges.

Let me know what you think 🤗🥳🤯

obmarg commented:
Thanks @carlocorradini but you really should have checked before going to the effort of putting this together.  

I don't have the time to reply to each individual change just now, but most of these changes are not things I want.

obmarg commented on Cargo.toml:
```diff
@@ -58,8 +65,24 @@ tokio-stream = { version = "0.1", features = ["sync"] }
+[workspace.lints.rust]
+unsafe_code = "forbid"
+
+[workspace.lints.clippy]
+pedantic = { level = "warn", priority = -1 }
+cargo = { level = "warn", priority = -1 }
+module_name_repetitions = "allow"
+multiple_crate_versions = "allow"
```
I am usually happy with clippys defaults.  I _definitely_ don't want `pedantic` warnings.

The cargo ones I'd be willing to consider, although `multiple_crate_versions` looks like one of the more useful ones.  What error is it giving if you don't have that allowed?

```suggestion
[workspace.lints.clippy]
cargo = { level = "warn", priority = -1 }
```

carlocorradini replied:
Fixed

obmarg commented on README.md:
```diff
@@ -1,20 +1,21 @@
   <h4>
-    <a href="https://github.com/obmarg/graphql-ws-client/tree/master/examples/examples">Examples</a>
+    <a href="https://github.com/obmarg/graphql-ws-client/tree/main/examples/examples">Examples</a>
     <span> | </span>
-    <a href="https://github.com/obmarg/graphql-ws-client/blob/master/CHANGELOG.md">Changelog</a>
+    <a href="https://github.com/obmarg/graphql-ws-client/blob/main/CHANGELOG.md">Changelog</a>
```
Thanks for finding these though, a good change - well spotted.

obmarg commented on tests/subscription_server/mod.rs:
```diff
@@ -116,6 +116,6 @@ pub struct SubscriptionRoot {
     async fn books(&self, _mutation_type: MutationType) -> impl Stream<Item = BookChanged> {
         println!("Subscription received");
-        BroadcastStream::new(self.channel.subscribe()).filter_map(|r| r.ok())
+        BroadcastStream::new(self.channel.subscribe()).filter_map(std::result::Result::ok)
```
I don't think this is an improvement tbh.

carlocorradini replied:
Clippy suggestion. Keep it or?

obmarg replied:
Whatever tbh

obmarg commented on src/ws_stream_wasm.rs:
```diff
@@ -29,6 +33,7 @@ impl crate::next::Connection for Connection {
     async fn receive(&mut self) -> Option<crate::next::Message> {
         use crate::next::Message;
         loop {
+            #[allow(clippy::match_same_arms)] // FIXME Remove when MSRV >= 1.78.0
```
Why "Remove when MSRV >= 1.78.0"?

carlocorradini replied:
Rust 1.76 (clippy) does not detect that the match arms have different meanings. With newer Rust versions this is fixed.

obmarg replied:
Can you just remove this please?  It's a pedantic lint and I do not want pedantic lints

obmarg commented on .gitignore:
```diff
@@ -1 +1,35 @@
+
+# Visual Studio Code
+/.vscode/*
+!/.vscode/settings.json
+!/.vscode/tasks.json
+!/.vscode/launch.json
+!/.vscode/extensions.json
+!/.vscode/*.code-snippets
```
I'd rather ignore all vscode stuff tbh

carlocorradini replied:
Done

obmarg commented on .gitignore:
```diff
@@ -1 +1,35 @@
+# Local History for Visual Studio Code
+/.history/
+
+# Built Visual Studio Code Extensions
+*.vsix
```
Not sure why we'd have any built vs code extensions?

obmarg reviewed (changes requested):
Sorry for the delay getting back to you @carlocorradini - been a busy week at work.

> * Added [Dependabot](https://github.com/features/security)
>   _Automatically monitor and fix vulnerable dependencies_

I'm planning to use renovate for this, see #90.  I just need to find the time to finish configuring it as I want.

> * Added [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)
>   _File to define individuals or teams that are responsible for code in a repository_

You can keep this, but I'm not sure what the purpose is.  It's my repo, obviously I am the owner.

> * Better CI/CD (see `.github/workflows`)

In what way is this better? I may be missing something, but it looks like you just rename the files?  If there are improvements I'm missing let me know, but I'd rather not rename the files without a good reason, so if you can revert that please.

> * Added [clippy](https://doc.rust-lang.org/clippy)
>   A collection of lints to catch common mistakes and improve your Rust code.
>   See `lints` in `Cargo.toml`
>   Some code has been refactored (**no breaking changes**) to fix all lints

I already use clippy on this repo, but I'm generally happy with its default settings.  I'll comment in `Cargo.toml` with more details though.

> * Added [.vscode](https://code.visualstudio.com) directory
>   * Added [extensions.json](https://code.visualstudio.com/docs/editor/extension-marketplace#_workspace-recommended-extensions) file
>     _Workspace recommended extensions_
>   * Added [settings.json](https://code.visualstudio.com/docs/getstarted/settings#_settingsjson)

I'd rather not dictate a users vscode settings - can you remove this please?

>     _Workspace settings_
> * Added [EditorConfig](https://editorconfig.org)
>   _maintain consistent coding styles for multiple developers working on the same project across various editors and IDEs_

I won't be using this myself so I'm not sure that there's any benefit to others using it?  It could easily not match the settings I've got.  I'm not that bothered about most of the settings in there either tbh - rustfmt deals with the majority of the formatting in the repo.

> * Added [.gitattributes](https://git-scm.com/docs/gitattributes)
>   _Defining attributes per path_

Is there anything particularly useful in this?  It's got more entries than there are files in this repository - and the vast majority of entries in this file are not likely to every be present in this repo.

> * Better `.gitignore`
> * `Cargo.lock` should be ignored in libraries (added in `.gitignore`)

I disagree, can you put the `Cargo.lock` back please.

> * Fixed some typos

Cool, thanks.

> * Better `README.md` (all `*.md` files in general) with better badges.

I don't think the formatting changes are an improvement, can you revert those.  I was also kind of fine with the badges as they were, what is better about these?

obmarg commented on .github/workflows/rust.yaml:
```diff
@@ -0,0 +1,51 @@
+                  components: clippy, rustfmt
+            - name: Check format
+              run: cargo fmt --all --check
+            - name: Check lint
+              run: cargo clippy --all --all-features
```
I'm probably fine with adding clippy to the existing rust workflow

obmarg commented on .github/workflows/rust.yaml:
```diff
@@ -0,0 +1,51 @@
+        runs-on: ubuntu-latest
+
+        steps:
+            - uses: actions/checkout@v4
+              with:
+                  fetch-depth: 0
+            - uses: dtolnay/rust-toolchain@1.76.0
+            - uses: Swatinem/rust-cache@v2
+            - name: Build
+              run: cargo build --workspace --all-features
+            - name: Examples
+              run: cargo build --workspace --all-features --examples
+            - name: Test
+              run: cargo test --workspace --all-features
```
But these changes are not ok. You have got rid of the separate test build step that was there for a reason, and also removed the example tests.

obmarg replied:
I am of course fine with renaming `--all` to `--workspace`

carlocorradini commented:
Sorry for the late reply; I'll fix them as soon as possible.

carlocorradini commented:
`.gitattributes` file is a common one that I use in all my projects and comes from the gitattributes repository.
The file is very useful to have a common/shared knowledge of line terminator to avoid possible commit changes across different environments (Widnows, Apple, Linux, ...)

carlocorradini commented:
`Cargo.lock` should NOT be committed.
[See official issue](https://github.com/rust-lang/cargo/issues/315).
Moreover, when you create a new library project the `.gitignore` file already contains `Cargo.lock` with a comment that it should be ignored if this is a library whereas maintained if this is a binary.

carlocorradini commented:
For all markdown files I've fixed some blanks that have been left and checked the formatting/styling with [markdownlint](https://github.com/DavidAnson/markdownlint) (i.e. added a blank between a title and text). Moreover, fixed `NewIssue` in `CONTRIBUTING.md`

obmarg commented:
> `Cargo.lock` should NOT be committed. [See official issue](https://github.com/rust-lang/cargo/issues/315). Moreover, when you create a new library project the `.gitignore` file already contains `Cargo.lock` with a comment that it should be ignored if this is a library whereas maintained if this is a binary.

First off - it doesn't really matter what a cargo issue from 2014 says.  It's my project and I want to commit `Cargo.lock`.

Secondly, this has not been the [official stance for coming up to a year](https://blog.rust-lang.org/2023/08/29/committing-lockfiles.html).

Third, that is not the behaviour for a new library project:

![CleanShot 2024-06-18 at 11 37 19@2x](https://github.com/obmarg/graphql-ws-client/assets/556490/4f1d5638-5373-484d-97c0-07814ae5b2aa)

carlocorradini commented:
@obmarg Awesome 🥳 
I didn't know about that blog post 🤗

carlocorradini commented:
@obmarg Everything should be fixed now 🤗🥳

obmarg commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,34 @@
+name: release
```
By renaming this file you've made it difficult for me to see what if anything has changed.  Can you put if back where it was please

obmarg commented on .github/workflows/rust.yaml:
```diff
@@ -0,0 +1,57 @@
+name: rust
```
By renaming this file you've made it difficult for me to see what if anything has changed. Can you put if back where it was please

obmarg commented on CHANGELOG.md:
```diff
@@ -15,263 +15,263 @@ all APIs might be changed.
 
-- Fixed some compile errors when `ws_stream_wasm` was enabled and `tungstenite`
-  was not ([#111](https://github.com/obmarg/graphql-ws-client/pull/111))
+-   Fixed some compile errors when `ws_stream_wasm` was enabled and `tungstenite`
```
Can you revert the formatting changes in this and all the other markdown files please.

obmarg commented on .gitattributes:
```diff
@@ -0,0 +1,221 @@
+* text=auto eol=lf
+
+# Source code
+*.bash text eol=lf
```
I've never personally run into problems working without a `.gitattributes`. I've looked around at a bunch of popular rust crates and can't see a `.gitattributes` file in those either.  Which leads me to believe this file is not that useful.  Can you remove it please.

carlocorradini commented:
@obmarg Done 😊

obmarg commented on .github/workflows/release.yml:
```diff
@@ -1,34 +1,34 @@
-          GITHUB_TOKEN: ${{ steps.generate-token.outputs.token }}
-          CARGO_REGISTRY_TOKEN: ${{ secrets.CARGO_REGISTRY_TOKEN }}
+    release-plz:
+        name: Release-plz
```
Can you revert all the indentation changes in these files as well please.

obmarg commented on .github/workflows/release.yml:
```diff
@@ -1,34 +1,34 @@
-name: release-plz
+name: release
```
Can you put this name back as well

obmarg commented on .github/workflows/rust.yml:
```diff
@@ -1,72 +1,57 @@
-        shell: bash
-        run: |
-          cargo test --all --examples
+    check:
```
Can you put this back to check-format please

obmarg commented on .github/workflows/rust.yml:
```diff
@@ -1,72 +1,57 @@
+                  components: clippy, rustfmt
+            - name: Check format
+              run: cargo fmt --all --check
+            - name: Check lint
+              run: cargo clippy --all --all-features
```
This job was specifically for checking formats.  I'm not against running clippy on CI, but it should be part of the build step so it can share build artifacts with all the other builds.

So can you move this into the build section please
